controllers/pulp/utils.go

```diff
@@ -245,3 +250,200 @@ func mountCASpec(pulp *repomanagerv1alpha1.Pulp, volumes []corev1.Volume, volume
 
 	return volumes, volumeMounts
 }
+
+// checkImmutableFields verifies if a user tried to modify an immutable field and rollback
+// the change if so
+func (r *PulpReconciler) checkImmutableFields(ctx context.Context, pulp *repomanagerv1alpha1.Pulp, field string, log logr.Logger) bool {
+
+	// access the field by its string name
+	// for fieldSpec we need to pass it as a reference because we will need to change
+	// its value back in case of immutable field
+	fieldSpec := reflect.Indirect(reflect.ValueOf(&pulp.Spec)).FieldByName(field)
+
+	// for fieldStatus, as we just need its content, we dont need to get a
+	// pointer to it
+	fieldStatus := reflect.ValueOf(pulp.Status).FieldByName(field)
+
+	// first we need to call the Interface() method to use the field as interface{}
+	// then we assert that the interface{} is a string so we can check if the len > 0
+	// if len>0 means that the field was previously defined
+	// after that we check if the content from .status.field is different from spec.field
+	// if so we should rollback the modification
+	if len(fieldStatus.Interface().(string)) > 0 && fieldStatus.Interface() != fieldSpec.Interface() {
+		patch := client.MergeFrom(pulp.DeepCopy())
+
+		// set pulp.spec.<field> back with the value from .status
+		fieldSpec.SetString(fieldStatus.Interface().(string))
+
+		// we are using patch here because we need to modify only a specific field.
+		// if we had used update it would fill a lot of other fields with default values
+		// which would also trigger a reconciliation loop
+		r.Patch(ctx, pulp, patch)
+		err := fmt.Errorf("%s field is immutable", field)
+		log.Error(err, "Could not update "+field+" field")
+		return true
+	}
+	return false
+}
+
+// updateIngressType will check the current definition of ingress_type and will handle the different
+// modification scenarios
+func (r *PulpReconciler) updateIngressType(ctx context.Context, pulp *repomanagerv1alpha1.Pulp) {
+
+	// if route => nodeport
+	// delete all routes with operator's labels
+	// remove route .status.conditions
+	// update .status.ingress_type = nodeport
+	if strings.ToLower(pulp.Status.IngressType) == "route" && strings.ToLower(pulp.Spec.IngressType) == "nodeport" {
```

Comment:
I think we should split this,
have 1 if for Status.IngressType == "route" and .Spec.IngressType != "route"
where we delete the routes,
and other if for specific nodeport needs

Reply:
hum... I'm not sure if I get your idea.
I let a "skeleton" with all of the possible combinations, for example:
* if it was _route_ and user modified to _ingress_ do A
* if it was _route_ and user modified to _nodeport_ do B
* if it was _route_ and user modified to _loadbalancer_ do C
* if it was _ingress_ and user modified to _route_ do D
... and the other possible combinations

to do that I'm using the same `.status.ingress_type` field to know what type it was before the user tried to modify pulp CR `.spec.ingress_type`:
* if `status.ingress_type=route` and  `.spec.ingress_type=ingress` do  A
* if `status.ingress_type=route` and  `.spec.ingress_type=nodeport` do  B
* if `status.ingress_type=route` and  `.spec.ingress_type=loadbalancer` do  C
* if `status.ingress_type=ingress` and  `.spec.ingress_type=route` do  D
... and the other possible combinations

Reply:
my idea is:
no matter what is the new ingress_type, if the old one was route, you need to delete them.

With the current logic you would repeat the deletion for each ingress type